lightdash/lightdash
# feat: [Custom dimension] Binning with fixed number

Closes: #5198

### Description



Allow users to define a dimension that is going to return a fixed amount of bins. 

Design mockups:
https://www.figma.com/file/JLRWyriHFvNnGu6F3klL8o/Untitled?node-id=0-1

- [x] only numeric dimensions can be 'binned'
- [x] if you have a numeric dimension, you can open the three-dot menu and see the option to `create custom dimension` (within this, there is the option to select `bin`)
- [x] There is a modal that pops up to create a binned custom dimension
- [x] You must name your custom dimension.
  - [x] By default, this is just `[name of dimension] bins` (e.g. creating a bin on a dimension `age` would be named `age bins` by default)
- [x] There are no `Bin Types` section (yet) - the only option is fixed number of bins for this ticket.
- [x] There is a number picker (like in filters) where you select the number of bins that you want. 
- [x] This number must be an integer.
- [ ] There is copy below the number of bins which tells you what the bin ranges will be for your data 
- [ ] There is copy that warns the user that these bins might change if their underlying data changes. 
- [ ] Once you've saved the custom dimension, it appears on the bottom of the sidebar under `custom dimensions`, above the `metrics` section, but below the `dimensions` section 
- [ ] The custom dimension is selected by default and added to your table. It behaves like a custom metric (for deletion and persistence)

Extra acceptance criteria: 

- [x] Make sure it works for charts
- [x] Sort bins
- [x] Make sure you can select this custom dimension on a chart
- [x] Chart tooltips 
- [x] Pie charts
- [x] Grouped charts
- [x] Pivot tables https://github.com/lightdash/lightdash/issues/7563
- [x] table charts
- [x] You can delete a custom dimension from explore 
- [x] It works with filters  https://github.com/lightdash/lightdash/issues/7564
- [x] It works with sorting
- [x] Build unit tests for bulidquery
- [x] Build e2e tests for custom dimension https://github.com/lightdash/lightdash/issues/7580
- [x] Disable `non-number` inputs in bins
- [ ] Work on `preview results` on the FE
- [ ] Rework `toggleActiveField`
- [x] Test underlying dasta - created a ticket
- [x] Wrap this in feature flag
- [x] Replace getCustomDimensionId with unique label toa void conflicts with other custom dimensions
- [x] Test multiple custom dimensions from the same dimension
- [x] Reference lines - created a ticket
- [x] CTE compatible with table calculations
- [x] persistence (migration + get + save) 
- [x] 1 bucket only edge case
- [x] multiple joins 
- [x] Test on dashboards 

Alternative SQL implementation https://github.com/lightdash/lightdash/pull/7545

Technical details:
1. add custom dimension to `MetricQuery` type. We should basically follow what we did for the custom metric ( aka additional metric ) 
2. if a binning custom dimension is selected, the query builder should:
  - add the "original" dimension in the main query
  - add the necessary sql in the second query to create the bins
  
 https://stackoverflow.com/questions/15464574/use-google-bigquery-to-build-histogram-graph
 

3. if we are sorting by the binning custom dimension, the query builder should:
  - add a custom order by 
```
// example
ORDER BY CASE `age`
         WHEN '< 0' THEN 1
         WHEN '0 to 100' THEN 2
         WHEN '100+' THEN 3
         ELSE 4
         END,
         `age`  
```
4. in the chart visualization, if the binning custom dimension is used, we need to add custom sorting to the series. Otherwise, it will be sorted alphabetically. 



<img width="900" alt="Captura de ecrã 2023-04-25, às 16 21 42" src="https://user-images.githubusercontent.com/9117144/234325278-e4bdc94b-1e1c-431c-a4e5-5481aabd4375.png">




### What problem does this solve?



As a user, I want to be able to create histograms so that I can see how my data is distributed across a specific number of groups.
### Reviewer actions
